dashboard/test/models/script_test.rb

```diff
@@ -1876,6 +1894,270 @@ class SummarizeVisibleAfterScriptTests < ActiveSupport::TestCase
+      stage 'Lesson1'
+      level '#{l.name}'
+    SCRIPT
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:stages]
+    )
+    assert_equal 'required', script.stages[0].lesson_group.key
+
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:stages]
+    )
+
+    assert_equal 'content', script.stages[0].lesson_group.key
+  end
+
+  test 'can remove the lesson group for a lesson' do
+    l = create :level
+    old_dsl = <<-SCRIPT
+      lesson_group 'required', display_name: 'Overview'
+      stage 'Lesson1'
+      level '#{l.name}'
+    SCRIPT
+    new_dsl = <<-SCRIPT
+      stage 'Lesson1'
+      level '#{l.name}'
+    SCRIPT
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:stages]
+    )
+
+    assert_equal 'required', script.stages[0].lesson_group.key
+
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:stages]
+    )
+
+    assert_equal '', script.stages[0].lesson_group.key
+  end
+
+  test 'can change the order of lesson groups in a script' do
+    l1 = create :level
+    l2 = create :level
+    old_dsl = <<-SCRIPT
+      lesson_group 'required', display_name: 'Overview'
+      stage 'Lesson1'
+      level '#{l1.name}'
+
+      lesson_group 'content', display_name: 'Content'
+      stage 'Lesson2'
+      level '#{l2.name}'
+    SCRIPT
+    new_dsl = <<-SCRIPT
+      lesson_group 'content', display_name: 'Content'
+      stage 'Lesson1'
+      level '#{l1.name}'
+
+      lesson_group 'required', display_name: 'Overview'
+      stage 'Lesson2'
+      level '#{l2.name}'
+    SCRIPT
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(old_dsl, 'a filename')[0][:stages]
+    )
+
+    assert_equal 'required', script.lesson_groups[0].key
+    assert_equal 1, script.lesson_groups[0].position
+    assert_equal 'content', script.lesson_groups[1].key
+    assert_equal 2, script.lesson_groups[1].position
+
+    script = Script.add_script(
+      {name: 'lesson-group-test-script'},
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:lesson_groups],
+      ScriptDSL.parse(new_dsl, 'a filename')[0][:stages]
+    )
+
+    assert_equal 'content', script.lesson_groups[0].key
+    assert_equal 1, script.lesson_groups[0].position
+    assert_equal 'required', script.lesson_groups[1].key
+    assert_equal 2, script.lesson_groups[1].position
+  end
+
```

Comment:
in this PR or the next one, please add a test for the following sequence:
1. a lesson group exists in the DSL and the DB 
2. the lesson group is removed from the DSL
3. add_script is run
EXPECTED: the lesson group is removed from the DB

For stages, this is super-sneakily accomplished by the following line: https://github.com/code-dot-org/code-dot-org/blob/f20739e5af13c69e68fe5376e00fbd5d50cfa0e2/dashboard/app/models/script.rb#L1104

by assigning the new stages, it's also un-assigning all of the old stages, causing the DB objects to be destroyed. A similar tactic could be used for lesson groups.

I'm not really sure how much of your other logic will have to be affected in order to catch this case, so if it is not super straight forward to address the earlier comments around find_or_create_by and get your existing tests passing again, then it may be less doing / undoing / redoing to add this test now. with the unique index on (script_id, key) and this new test, I think you should be in good shape to rely on tests passing as an indicator as to whether anything's being overlooked in the code.

Reply:
I added code similar to the example you gave for stages to delete old lesson groups. I also added one line at the end of the test `can remove the lesson group for a lesson` to make sure that we only have one lesson group for that scripts after the lesson group is changed which I think tests what you are asking for although it could be more explicit if you think that would be good.

Reply:
yes, this looks good!